cardano-node/src/Cardano/Tracing/Kernel.hs

```diff
@@ -0,0 +1,59 @@
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE DeriveGeneric         #-}
+{-# LANGUAGE DeriveTraversable     #-}
+{-# LANGUAGE FlexibleInstances     #-}
+
+module Cardano.Tracing.Kernel
+  ( NodeKernelData (..)
+  , mkNodeKernelData
+  , setNodeKernel
+  , mapNodeKernelDataIO
+  , SMaybe (..)
+  , fromSMaybe
+  -- * Re-exports
+  , NodeKernel (..)
+  , LocalConnectionId
+  , RemoteConnectionId
+  ) where
+
+import           Cardano.Prelude hiding (atomically)
+
+import           Data.IORef (IORef, newIORef, readIORef, writeIORef)
+
+import           Ouroboros.Consensus.Node (NodeKernel(..))
+import           Ouroboros.Consensus.Util.Orphans ()
+
+import           Ouroboros.Network.NodeToClient (LocalConnectionId)
+import           Ouroboros.Network.NodeToNode (RemoteConnectionId)
+
+
+data SMaybe a
+  = SNothing
+  | SJust !a
+  deriving (Foldable, Functor, Generic, NFData, NoUnexpectedThunks, Traversable)
+
+fromSMaybe :: a -> SMaybe a -> a
+fromSMaybe x SNothing = x
+fromSMaybe _ (SJust x) = x
+
+
+newtype NodeKernelData blk =
+  NodeKernelData
+  { unNodeKernelData :: IORef (SMaybe (NodeKernel IO RemoteConnectionId LocalConnectionId blk))
```

Comment:
This `IORef` will only be filled in once, when the `NodeKernel` has started (I know that was already the case before this PR), so that `IORef` + `SMaybe` is only used to break the circular dependency right?

Reply:
Yes, the tracers have to be created to make `NodeKernel` available.